commit bcf11809b68e8c97585955e55ea2a935c7d2d1fe (HEAD)
Author: Ken Zhu <quic_guigenz@quicinc.com>
Date:   Tue May 7 18:10:22 2024 -0700

    netifd: add no hairpin flood ports configuration

    add no hairpin flood ports option in the device bridge.

    Change-Id: Id09a52225ee8ca5f6f3591b48dfe5a1d0dfcd2e6
    Signed-off-by: Ken Zhu <quic_guigenz@quicinc.com>

diff -Nur a/bridge.c b/bridge.c
--- a/bridge.c	2024-05-06 15:23:59.878712238 -0700
+++ b/bridge.c	2024-05-06 19:40:15.086289959 -0700
@@ -16,6 +16,7 @@
 #include <stdio.h>
 #include <assert.h>
 #include <errno.h>
+#include <ctype.h>
 
 #include "netifd.h"
 #include "device.h"
@@ -44,6 +45,7 @@
 	BRIDGE_ATTR_HAS_VLANS,
 	BRIDGE_ATTR_STP_KERNEL,
 	BRIDGE_ATTR_STP_PROTO,
+	BRIDGE_ATTR_NO_FLOOD_PORTS,
 	__BRIDGE_ATTR_MAX
 };
 
@@ -68,10 +70,12 @@
 	[BRIDGE_ATTR_HAS_VLANS] = { "__has_vlans", BLOBMSG_TYPE_BOOL }, /* internal */
 	[BRIDGE_ATTR_STP_KERNEL] = { "stp_kernel", BLOBMSG_TYPE_BOOL },
 	[BRIDGE_ATTR_STP_PROTO] = { "stp_proto", BLOBMSG_TYPE_STRING },
+	[BRIDGE_ATTR_NO_FLOOD_PORTS] = { "no_flood_ports", BLOBMSG_TYPE_ARRAY },
 };
 
 static const struct uci_blob_param_info bridge_attr_info[__BRIDGE_ATTR_MAX] = {
 	[BRIDGE_ATTR_PORTS] = { .type = BLOBMSG_TYPE_STRING },
+	[BRIDGE_ATTR_NO_FLOOD_PORTS] = { .type = BLOBMSG_TYPE_STRING },
 };
 
 static const struct uci_blob_param_list bridge_attr_list = {
@@ -117,6 +121,7 @@
 	struct blob_attr *config_data;
 	struct bridge_config config;
 	struct blob_attr *ports;
+	struct blob_attr *no_flood_ports;
 	bool active;
 	bool force_active;
 	bool has_vlans;
@@ -135,6 +140,7 @@
 	struct uloop_timeout check_timer;
 	struct device_vlan_range *extra_vlan;
 	int n_extra_vlan;
+	int no_flood;
 	uint16_t pvid;
 	bool present;
 	bool active;
@@ -592,6 +598,31 @@
 }
 
 static int
+bridge_is_no_flood_dev(struct blob_attr *no_flood_ports, struct device *dev)
+{
+	struct blob_attr *cur;
+	size_t rem;
+	char *ifname;
+	char vlan_base[IFNAMSIZ];
+	char *dot;
+	dot = strchr(dev->ifname, '.');
+
+	if (dot && isdigit(*(dot + 1))) {
+		size_t len = dot - dev->ifname;
+		strlcpy(vlan_base, dev->ifname,  len - 1);
+		ifname = vlan_base;
+	} else {
+		ifname = dev->ifname;
+	}
+
+	blobmsg_for_each_attr(cur, no_flood_ports, rem) {
+		if (strcmp(blobmsg_data(cur), ifname) == 0)
+			return 1;
+	}
+	return 0;
+}
+
+static int
 bridge_enable_member(struct bridge_member *bm)
 {
 	struct bridge_state *bst = bm->bst;
@@ -641,6 +672,10 @@
 		system_bridge_set_hairpin(bm->dev.dev, bst->config.hairpin);
 	}
 
+	if (bm->no_flood) {
+		system_bridge_set_no_hairpin_flood(bm->dev.dev, "1");
+	}
+
 	if (!dev->settings.auth || dev->auth_status)
 		device_broadcast_event(&bst->dev, DEV_EVENT_TOPO_CHANGE);
 
@@ -894,6 +929,7 @@
 	bm->check_timer.cb = bridge_member_check_cb;
 	strcpy(bm->name, name);
 	bm->dev.dev = dev;
+	bm->no_flood = bridge_is_no_flood_dev(bst->no_flood_ports, dev);
 
 	return bm;
 }
@@ -1098,6 +1134,17 @@
 
 	blobmsg_close_array(b, list);
 
+	list = blobmsg_open_array(b, "no_flood_ports");
+
+	vlist_for_each_element(&bst->members, bm, node) {
+		if (!bm->no_flood)
+			continue;
+
+		blobmsg_add_string(b, NULL, bm->dev.dev->ifname);
+	}
+
+	blobmsg_close_array(b, list);
+
 	c = blobmsg_open_table(b, "bridge-attributes");
 
 	blobmsg_add_u8(b, "stp", cfg->stp);
@@ -1279,6 +1326,7 @@
 		bst->primary_port = NULL;
 
 	bst->ports = tb_br[BRIDGE_ATTR_PORTS];
+	bst->no_flood_ports = tb_br[BRIDGE_ATTR_NO_FLOOD_PORTS];
 	device_init_settings(dev, tb_dev);
 	bridge_apply_settings(bst, tb_br);
 
diff -Nur a/system.h b/system.h
--- a/system.h	2024-05-06 15:23:59.882712305 -0700
+++ b/system.h	2024-05-06 19:40:27.922503018 -0700
@@ -249,6 +249,7 @@
 int system_bridge_addif(struct device *bridge, struct device *dev);
 int system_bridge_delif(struct device *bridge, struct device *dev);
 void system_bridge_set_hairpin(struct device *dev, int enable);
+void system_bridge_set_no_hairpin_flood(struct device *dev, const char *val);
 int system_bridge_vlan(const char *iface, uint16_t vid, int16_t vid_end, bool add, unsigned int vflags);
 int system_bridge_vlan_check(struct device *dev, char *ifname);
 void system_bridge_set_stp_state(struct device *dev, bool val);
diff -Nur a/system-linux.c b/system-linux.c
--- a/system-linux.c	2024-05-06 15:23:59.878712238 -0700
+++ b/system-linux.c	2024-05-06 19:43:32.337563773 -0700
@@ -577,6 +577,11 @@
 	system_set_dev_sysfs("brport/unicast_flood", dev->ifname, val);
 }
 
+void system_bridge_set_no_hairpin_flood(struct device *dev, const char *val)
+{
+	system_set_dev_sysfs("brport/no_hairpin_flood", dev->ifname, val);
+}
+
 static int system_get_disable_ipv6(struct device *dev, char *buf, const size_t buf_sz)
 {
 	return system_get_dev_sysctl("ipv6/conf", "disable_ipv6",
